File: stats.py
from flask import Flask, jsonify, render_template, request
from data import *

def stats():
    return render_template("stats.html", champs=champs)

def query():
    # Do the thing according to filters
    filters = request.get_json("filters")
    gameNumber = int(filters["gameNumber"])
    filters.pop("gameNumber")
    games = Game.query
    pop = []
    for filter in filters:
        if not filters[filter]:
            pop.append(filter)
    for filter in pop:
        filters.pop(filter)
    games = games.filter_by(**filters)
    games = games.order_by(Game.id.desc()).limit(gameNumber).all()
    if games:
        # Defining and resetting variables
        games_played = len(games)
        game_modes = {"normal": 0, "ranked": 0, "flex": 0, "aram": 0}
        roles = {"top": 0, "jg": 0, "mid": 0, "adc": 0, "sup": 0}
        wins = 0
        ka = 0
        d = 0
        cs = 0
        gold = 0
        vision = 0
        duration = 0

        # Iterate through games to create results
        for match in games:
            mode = match.game_mode
            role = match.role
            game_modes[mode] += 1
            if role != "":
                roles[role] += 1
            if match.result == "W":
                wins += 1
            ka += match.k + match.a
            d += match.d
            cs += match.cs
            gold += match.gold
            duration += match.duration
            if match.vision != "":
                vision += int(match.vision)

        # Work with the results
        kda = round(ka / d, 2)
        winrate = round(wins * 100 / games_played, 2)
        csm = round(cs * 60 / duration, 2)
        gpm = round(gold * 60 / duration, 2)
        vision = round(vision * 3600 / duration, 2)
        avg_duration = int(duration / games_played)
        avg_duration = f"{avg_duration // 60:02}:{avg_duration % 60:02}"

        # Compose the JSON to return
        data = {
            "success": True,
            "gameModes": game_modes,
            "roles": roles,
            "gamesPlayed": games_played,
            "winrate": winrate,
            "kda": kda,
            "csm": csm,
            "gpm": gpm,
            "vision": vision,
            "averageDuration": avg_duration
        }
        return jsonify(data)
    else:
        return jsonify({"success": False})

File: static/js/stats.js

    // Load google charts
    function loadCharts() {
        google.charts.load('current', {'packages':['corechart']})
        google.charts.setOnLoadCallback(drawCharts)
    }

    // Draw the chart and set the chart values
    function drawCharts() {

        var gameModes = google.visualization.arrayToDataTable([
        ['Mode', 'Games played'],
        ['Normal', parseInt(document.getElementById('normal').innerHTML)],
        ['Ranked', parseInt(document.getElementById('ranked').innerHTML)],
        ['Flex', parseInt(document.getElementById('flex').innerHTML)],
        ['ARAM', parseInt(document.getElementById('aram').innerHTML)]
    ]);

        var role = google.visualization.arrayToDataTable([
        ['Role', 'Games played'],
        ['Top', parseInt(document.getElementById('top').innerHTML)],
        ['JG', parseInt(document.getElementById('jg').innerHTML)],
        ['Mid', parseInt(document.getElementById('mid').innerHTML)],
        ['ADC', parseInt(document.getElementById('adc').innerHTML)],
        ['Support', parseInt(document.getElementById('sup').innerHTML)],
    ]);

        // Optional; add a title and set the width and height of the chart
        var gameModesTitle = {'title':'Game modes', 'width': 500, 'height': 300};
        var roleTitle = {'title': 'Role distribution', 'width': 500, 'height': 300}

        // Display the chart inside the <div> element with id="piechart"
        var piechart1 = new google.visualization.PieChart(document.getElementById('piechart1'));
        var piechart2 = new google.visualization.PieChart(document.getElementById('piechart2'));
        piechart1.draw(gameModes, gameModesTitle);
        piechart2.draw(role, roleTitle);
    }

    function updateData() {
        // Create AJAX request and define values
        const request = new XMLHttpRequest(),
        filters = {
        "game_mode": document.querySelector('#mode').value != '' ? document.querySelector('#mode').value : null,
        "champ": document.querySelector('#champ').value != '' ? document.querySelector('#champ').value : null,
        "role": document.querySelector('#role').value != '' ? document.querySelector('#role').value : null,
        "gameNumber": document.querySelector('#number').value
        }

        // Callback function for when request completes
        request.onreadystatechange = () => {
            // Extract data
            if (request.readyState == 4 && request.status == 200) {
            const data = JSON.parse(request.responseText)

            // Update HTML
            if (data.success === true) {
                document.getElementById('normal').innerHTML = data.gameModes.normal
                document.getElementById('ranked').innerHTML = data.gameModes.ranked
                document.getElementById('flex').innerHTML = data.gameModes.flex
                document.getElementById('aram').innerHTML = data.gameModes.aram
                document.getElementById('top').innerHTML = data.roles.top
                document.getElementById('jg').innerHTML = data.roles.jg
                document.getElementById('mid').innerHTML = data.roles.mid
                document.getElementById('adc').innerHTML = data.roles.adc
                document.getElementById('sup').innerHTML = data.roles.sup
                document.getElementById('games').innerHTML = data.gamesPlayed
                document.getElementById('winrate').innerHTML = data.winrate
                document.getElementById('kda').innerHTML = data.kda
                document.getElementById('csm').innerHTML = data.csm
                document.getElementById('gpm').innerHTML = data.gpm
                document.getElementById('duration').innerHTML = data.averageDuration
                document.getElementById('vision').innerHTML = data.vision
                loadCharts()
            }
            else {
                console.log("No games found")
            }
            }
        }
        request.open('POST', '/query')
        request.send(JSON.stringify(filters))
    }
document.addEventListener('DOMContentLoaded', () => {
    document.getElementById('mode').setAttribute('onchange', 'updateData()')
    document.getElementById('champ').setAttribute('onchange', 'updateData()')
    document.getElementById('role').setAttribute('onchange', 'updateData()')
    document.getElementById('number').setAttribute('onchange', 'updateData()')
    updateData()
})